// @material-extended/mde/esm2015/lib/popover/popover-interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function MdePopoverPanel() { }
function MdePopoverPanel_tsickle_Closure_declarations() {
    /** @type {?} */
    MdePopoverPanel.prototype.positionX;
    /** @type {?} */
    MdePopoverPanel.prototype.positionY;
    /** @type {?} */
    MdePopoverPanel.prototype.containerPositioning;
    /** @type {?} */
    MdePopoverPanel.prototype.overlapTrigger;
    /** @type {?} */
    MdePopoverPanel.prototype.triggerEvent;
    /** @type {?} */
    MdePopoverPanel.prototype.enterDelay;
    /** @type {?} */
    MdePopoverPanel.prototype.leaveDelay;
    /** @type {?} */
    MdePopoverPanel.prototype.targetOffsetX;
    /** @type {?} */
    MdePopoverPanel.prototype.targetOffsetY;
    /** @type {?} */
    MdePopoverPanel.prototype.arrowOffsetX;
    /** @type {?} */
    MdePopoverPanel.prototype.arrowWidth;
    /** @type {?} */
    MdePopoverPanel.prototype.arrowColor;
    /** @type {?} */
    MdePopoverPanel.prototype.closeOnClick;
    /** @type {?} */
    MdePopoverPanel.prototype.closeDisabled;
    /** @type {?} */
    MdePopoverPanel.prototype.setCurrentStyles;
    /** @type {?} */
    MdePopoverPanel.prototype.templateRef;
    /** @type {?} */
    MdePopoverPanel.prototype.close;
    /** @type {?} */
    MdePopoverPanel.prototype.setPositionClasses;
    /** @type {?} */
    MdePopoverPanel.prototype._emitCloseEvent;
}
/**
 * @record
 */
export function MdePopoverConfig() { }
function MdePopoverConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    MdePopoverConfig.prototype.positionX;
    /** @type {?} */
    MdePopoverConfig.prototype.positionY;
    /** @type {?} */
    MdePopoverConfig.prototype.overlapTrigger;
    /** @type {?} */
    MdePopoverConfig.prototype.triggerEvent;
    /** @type {?} */
    MdePopoverConfig.prototype.triggerDelay;
    /** @type {?} */
    MdePopoverConfig.prototype.targetOffsetX;
    /** @type {?} */
    MdePopoverConfig.prototype.targetOffsetY;
    /** @type {?} */
    MdePopoverConfig.prototype.arrowOffsetX;
    /** @type {?} */
    MdePopoverConfig.prototype.arrowWidth;
    /** @type {?} */
    MdePopoverConfig.prototype.arrowColor;
    /** @type {?} */
    MdePopoverConfig.prototype.closeOnClick;
}
/**
 * @record
 */
export function MdeTarget() { }
function MdeTarget_tsickle_Closure_declarations() {
    /** @type {?} */
    MdeTarget.prototype._elementRef;
}
